### DISCUSSION TITLE
Welcome to m*lib Discussions!

### DISCUSSION THREAD

**Discussion by P-p-H-d** [P-p-H-d/mlib#91] [General]
## Welcome!
  Discussions have been opened as a place to communicate with other members for:
  * Asking questions you’re wondering about.
  * Sharing ideas.
  * Engaging with other community members.
  * Welcoming others.

**Comment by l-t-m-f** [comment]
I began using this library for an experimental game in C99. I like it so far. The hardest thing with it is just specifying the correct OPLIST, but now I'm starting to become more friendly with the system. 

I wanted to know if "m-shared.h" is still the right shared pointer header to use. I also was wondering which features are C11-only and if there is a list of them somewhere so I avoid them.

Finally, the library creates some false positive with Jetbrains new Nova Engine for C/C++ so I should let you know I have signaled the bug to the CLion team using an example from the library, to which they have replied and assigned tracking.

Thank you,

**Reply by P-p-H-d** [reply]
Hi,

First, thank you for using this library!

I know the OPLIST feature is really unique and may be harder to use than alternative. But the more I use it, the more I realize how powerful it is, allowing to create more and more features, some of them not even possible in classic C++! (like serialization, or API adaptation). I want to make it easier to use, and that come from having better compilation error. If you have any idea on how to misuse them, I'm in: I want to analyze how I can provide better error message in this case.

The header m-shared as been very recently deprecated by the header m-shared-ptr. I didn't like the old API, so this header provides a complete new API. I find it much cleaner and easy to use. The old interface doesn't feel like pointers when using it. The new interface provide a clear pointer object to the user. Moreover it is packed with more feature, enabling to translate any container into a concurrent container (like the old m-concurrent header) and enabling to separate declaration of functions and their definition (enabling proper encapsulation).  I'm open to any comment on it.

The features specific to C11 are linked to the _Generic keyword:

* string functions accepting either a string_t or a const char * as second argument. For example, iIn C99, you need to call either string_set or string_set_str. In C11, you can call   string_set and it will be mapped to the right function.
* serialization of integer / floats using m-serial headers is only available in C11. In C99, the user needs to write its own serialization function for each integer and float he wants to use (and providing specialized oplist instead of M_BASIC_OPLIST)
* better and more explicit error messages.

I suppose the bug you reported  is this one https://youtrack.jetbrains.com/issue/CPP-42348 ?

By the way, which game are you working on?

**Reply by l-t-m-f** [reply]
Hi, sorry for taking a while to respond, I don't have a good excuse! Your comment was helpful, I'm glad most of the API is C99 compatible. 

The issue was indeed CPP-42348, but Jetbrains notifed me recently that it was fixed. However, I verified the lastest CLion release of december 19th, and it didn't seem like it was the case. Maybe they mean their early access build, I have yet to check it out.

The game is my own personal project, its a 2D with isometric graphics. I'm using flecs and SDL3 with your library as well for containers and the string type. I don't have much to share about it. Since I'm basically building the whole thing from scratch it's moving slowly but surely. Its going to be an RPG with a vibe like Earthbound.
